python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -21,177 +21,842 @@
-                name for name in self.file.metadata.schema.names if not name.startswith("__")
+    extension = ".parq"
+
+    def read(self, component: Optional[str] = None) -> Any:
+        # Docstring inherited from Formatter.read.
+        schema = pq.read_schema(self.fileDescriptor.location.path)
+
+        if component in ("columns", "schema"):
+            # The schema will be translated to column format
+            # depending on the input type.
+            return schema
+        elif component == "rowcount":
+            # Get the rowcount from the metadata if possible, otherwise count.
+            if b"lsst::arrow::rowcount" in schema.metadata:
+                return int(schema.metadata[b"lsst::arrow::rowcount"])
+
+            temp_table = pq.read_table(
+                self.fileDescriptor.location.path,
+                columns=[schema.names[0]],
+                use_threads=False,
+                use_pandas_metadata=False,
             )
+
+            return len(temp_table[schema.names[0]])
+
+        par_columns = None
+        filters = None
+        if self.fileDescriptor.parameters:
+            par_columns = self.fileDescriptor.parameters.pop("columns", None)
+            if par_columns:
+                has_pandas_multi_index = False
+                if b"pandas" in schema.metadata:
+                    md = json.loads(schema.metadata[b"pandas"])
+                    if len(md["column_indexes"]) > 1:
+                        has_pandas_multi_index = True
+
+                if not has_pandas_multi_index:
+                    # Ensure uniqueness, keeping order.
+                    par_columns = list(dict.fromkeys(ensure_iterable(par_columns)))
+                    file_columns = [name for name in schema.names if not name.startswith("__")]
+
+                    for par_column in par_columns:
+                        if par_column not in file_columns:
+                            raise ValueError(
+                                f"Column {par_column} specified in parameters not available in parquet file."
+                            )
+                else:
+                    par_columns = _standardize_multi_index_columns(schema, par_columns)
+
+            filters = self.fileDescriptor.parameters.pop("filters", None)
+
+            if len(self.fileDescriptor.parameters):
+                raise ValueError(
+                    f"Unsupported parameters {self.fileDescriptor.parameters} in ArrowTable read."
+                )
+
+        metadata = schema.metadata if schema.metadata is not None else {}
+        arrow_table = pq.read_table(
+            self.fileDescriptor.location.path,
+            columns=par_columns,
+            use_threads=False,
+            use_pandas_metadata=(b"pandas" in metadata),
+            filters=filters,
+        )
+
+        return arrow_table
+
+    def write(self, inMemoryDataset: Any) -> None:
+        import numpy as np
+        import pandas as pd
```

Comment:
Importing pandas just to check if we have been given a pandas dataframe is a tad annoying.